## Projeto SQL

### Objetivos: 
Extrair informações relevantes sobre o serviço para gerar um novo produto. 

In [1]:
#conectando a base de dados: código copiado do notion conforme instruções
# importar librerías
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'practicum_student',         # nombre de usuario
             'pwd': 's65BlTKV3faNIGhmvJVzOqhs', # contraseña
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # puerto de conexión
             'db': 'data-analyst-final-project-db'}          # nombre de la base de datos

In [3]:
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

In [4]:
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [6]:
#importando tabelas
books = pd.read_sql('SELECT * FROM books', engine)
authors = pd.read_sql('SELECT * FROM authors', engine)
publishers = pd.read_sql('SELECT * FROM publishers', engine)
ratings = pd.read_sql('SELECT * FROM ratings', engine)
reviews = pd.read_sql('SELECT * FROM reviews', engine)

In [7]:
#estudando as tabelas
print('Primeiras 5 linhas:\n', books.head())

Primeiras 5 linhas:
    book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  


In [8]:
print('Primeiras 5 linhas:\n', authors.head())

Primeiras 5 linhas:
    author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


In [9]:
print('Primeiras 5 linhas:\n', publishers.head())

Primeiras 5 linhas:
    publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company


In [10]:
print('Primeiras 5 linhas:\n', ratings.head())

Primeiras 5 linhas:
    rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


In [11]:
print('Primeiras 5 linhas:\n', reviews.head())

Primeiras 5 linhas:
    review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  


#### Comentário: 
Os dados parecem estar corretos, pelo diagrama de dados disponível conferi os tipos de dados e estão corretos  também.

#### Tarefas:

In [12]:
#Encontre o número de livros lançados depois de 1 de janeiro de 2000.
query1 = """
SELECT 
    COUNT(DISTINCT book_id) as total_livros
FROM
    books
WHERE
    publication_date > '2000-01-01';
"""

resultado1 = pd.read_sql(query1, engine)

print("=== TAREFA 1 ===")
print("Número de livros lançados depois de 1º de janeiro de 2000:")
print(resultado1)


=== TAREFA 1 ===
Número de livros lançados depois de 1º de janeiro de 2000:
   total_livros
0           819


In [19]:
#Encontre o número de avaliações e a classificação média para cada livro.
query2 = """
SELECT 
    r.book_id,
    b.title,
    COUNT(r.rating) as total_avaliacoes,
    AVG(r.rating) as classificacao_media
FROM ratings r
JOIN books b ON r.book_id = b.book_id
GROUP BY r.book_id, b.title
ORDER BY total_avaliacoes DESC;
"""

resultado2 = pd.read_sql(query2, engine)

print("=== TAREFA 2 ===")
print("Número de avaliações e classificação média para cada livro:")
print(resultado2)

=== TAREFA 2 ===
Número de avaliações e classificação média para cada livro:
     book_id                                              title  \
0        948                            Twilight (Twilight  #1)   
1        750                The Hobbit  or There and Back Again   
2        673                             The Catcher in the Rye   
3         75               Angels & Demons (Robert Langdon  #1)   
4        302  Harry Potter and the Prisoner of Azkaban (Harr...   
..       ...                                                ...   
995      609           Swann's Way (In Search of Lost Time  #1)   
996      412                   Lucy Sullivan Is Getting Married   
997      635                                  The Art of Loving   
998      817                                    The Other Woman   
999      191  Disney's Beauty and the Beast (A Little Golden...   

     total_avaliacoes  classificacao_media  
0                 160             3.662500  
1                  88       

#### Comentário:
O total das avaliações variam bastante, indo de 1 até 160 por exemplo, já a classificação média se mantém na média de 3 a 4,5, sendo maior que isso é um livro de destaque e menor um livro não tão aclamado pelo público. Temos como Top1 de avaliações o livro 'Twilight' (crepúsculo).

In [20]:
#Identifique a editora que lançou o maior número de livros com mais de 50 páginas 
query3 = """
SELECT
    p.publisher, 
    COUNT(*) as total_livros
FROM
    books b
JOIN publishers p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY total_livros DESC
LIMIT 1;
"""

resultado3 = pd.read_sql(query3, engine)

print("=== TAREFA 3 ===")
print("Editora que lançou o maior número de livros com mais de 50 páginas:")
print(resultado3)

=== TAREFA 3 ===
Editora que lançou o maior número de livros com mais de 50 páginas:
       publisher  total_livros
0  Penguin Books            42


#### Comentário: 
A editora 'Penguin Books' foi a editora que lançou mais livros com mais de 50 páginas!

In [24]:
#Identifique o autor com a média mais alta de classificação de livros: olhe apenas para livros com pelo menos 50 classificações.
query4 = """
WITH book_ratings AS (
    SELECT
        b.book_id,
        b.author_id,
        AVG(r.rating) AS avg_rating,
        COUNT(r.rating_id) AS num_ratings
    FROM books AS b
    JOIN ratings AS r ON b.book_id = r.book_id
    GROUP BY b.book_id, b.author_id
    HAVING COUNT(r.rating_id) >= 50
)
    

SELECT
    br.author_id,
    a.author,
    AVG(br.avg_rating) AS author_avg_rating
FROM book_ratings AS br
JOIN authors AS a ON br.author_id = a.author_id
GROUP BY br.author_id, a.author
ORDER BY author_avg_rating DESC
LIMIT 1;

"""
resultado4 = pd.read_sql(query4, engine)
print("=== TAREFA 4 ===")
print("Autor com a média mais alta de classificação de livros (mínimo 50 classificações):")
print(resultado4)

=== TAREFA 4 ===
Autor com a média mais alta de classificação de livros (mínimo 50 classificações):
   author_id                      author  author_avg_rating
0        236  J.K. Rowling/Mary GrandPré           4.283844


#### Comentário:
A autora J.K Rowling, famosa pela saga de Harry Potter, é a que tem a média mais alta de classificações de livros, qual será o gênero que ela mais escreve? Seria uma métrica boa para avaliarmos. Junto com ela temos a Mary GrandPré, será ela uma co-escritora ou parceira? 

In [27]:
#Encontre o número médio de avaliações entre usuários que classificaram mais do que 50 livros.
query5 = """
SELECT
    AVG(review_count) AS media_livros_classificados
FROM (
    SELECT 
        COUNT(DISTINCT book_id) AS review_count
    FROM 
        reviews
    WHERE 
        username IN(
            SELECT
                username
            FROM
                ratings
            GROUP BY 
                username
            HAVING
                COUNT(DISTINCT book_id) > 50
        )
    GROUP BY username
) AS reviews_counts;

"""
resultado5 = pd.read_sql(query5, engine)
print("=== TAREFA 5 ===")
print("Número médio de avaliações entre usuários que classificaram mais do que 50 livros:")
print(resultado5.round(2))

=== TAREFA 5 ===
Número médio de avaliações entre usuários que classificaram mais do que 50 livros:
   media_livros_classificados
0                       24.33


#### Comentário:
A média de avaliações entre os usuários que classificaram mais do que 50 livros é de: 24 avaliações

### Conclusão: 
A média de avaliações realizadas entre os usuários que classificam mais do que 50 livros é cerca de 24 avaliações, onde a classificação média dos livros é de 3 a 4,5 pontos. A editora com mais livros publicados, com mais de 50 páginas, é a Penguim books com cerca de 42 livros e a autora com melhor classificação é a J.K Rowling e a Mary GrandPré com uma avaliação de 4,3 pontos. Podemos pensar em incluir mais livros dessa autora e em especial livros que ela publicou pela editora Penguim books. Também seria interessante avaliarmos os gêneros que a editora mais publica e que a autora mais escreve, assim poderiamos explorar mais livros desse gênero. E o livro com mais avaliações é o famoso Twilight (Crepúsculo).